---
title: "Human and murine responses to Mtb"
subtitle: "Ortholog genes in IFN responses"
author: "Kim Dill-McFarland, kadm@uw.edu"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: no
date: "version `r format(Sys.time(), '%B %d, %Y')`"
editor_options: 
  chunk_output_type: console
---
# Setup
Load packages

```{r message=FALSE, warning=FALSE}
# Data manipulation and figures
library(tidyverse)
library(janitor)
library(readxl)
# Broad db
library(msigdbr)
# library(BIGpicture)
library(patchwork)
# library(readxl)
#Print pretty tables to Rmd
# library(DT)
```

Set seed

```{r}
set.seed(589)
```

# Load data

```{r data}
hum_FC <- read_csv("data_raw/human/AM-MDM/AM-MDM_gene_pval.csv") %>% 
  select(group, geneName, hgnc_symbol, logFC, adj.P.Val) %>% 
  rename(ensemblID = geneName, symbol = hgnc_symbol, FDR = adj.P.Val) %>% 
  filter(group %in% c("AM_TB - AM_Media", "MDM_TB - MDM_Media")) %>% 
  separate(group, into = c("group"), extra = "drop") %>% 
  mutate(group = paste("Human", group))
```

```{r}
mur_mdm <- read_excel("data_raw/murine/BMDM_H37Rv_RNAseq.xlsx", sheet = "all genes") %>% 
  rename(ensemblID = Geneid, symbol = gene, logFC = gso30.Bl6.N.dRv.4.logFC, 
         FDR = gso30.Bl6.N.dRv.4.FDR) %>% 
  mutate(group = "Murine BMDM")

mur_am <- read_excel("data_raw/murine/AM_H37Rv_RNAseq.xlsx", sheet = "Fit coefficients") %>% 
  pivot_longer(-c(Geneid, gene)) %>% 
  filter(grepl("^d.", name)) %>% 
  separate(name, into = c("trash","group","trash2","name"), extra = "drop") %>% 
  pivot_wider() %>% 
  rename(ensemblID = Geneid, symbol = gene) %>% 
  select(-trash, -trash2) %>% 
  mutate(group = recode(group, "unvax"="Murine AM",
                        "scBCG" = "Murine AM scBCG",
                        "cotb" = "Murine AM coTB"))
#Combine
mur_FC <- bind_rows(mur_mdm, mur_am) %>% 
  mutate(symbol = toupper(symbol))
```

Join human and mouse data.

```{r}
dat <- bind_rows(hum_FC, mur_FC) %>% 
  mutate(group = factor(group)) %>% 
  mutate(Significance = case_when(FDR < 0.01 ~ "FDR < 0.01",
                                  FDR < 0.01 ~ "FDR < 0.05",
                                  FDR < 0.01 ~ "FDR < 0.1",
                                  FDR < 0.01 ~ "FDR < 0.2")) %>% 
  arrange(symbol, group)
```

Check and recode when gene symbols differ. 
**Check all IFN in db
**Note IFN with "messy" ortho anno
**IFNb gene set
**human vs BCG all genes in diff directions
**fine scrnaseq tb data sets
```{r}
key_raw <- read_tsv("data_raw/Jax_HOM_AllOrganism.rpt") %>% 
  clean_names() %>% 
  filter(common_organism_name %in% c("human","mouse, laboratory")) %>% 
  mutate(common_organism_name = recode(common_organism_name, "mouse, laboratory"="mouse"),
         symbol = toupper(symbol)) %>% 
  distinct(db_class_key, common_organism_name, symbol) %>% 
  pivot_wider(names_from = common_organism_name, values_from = symbol,
              values_fn = list) %>% 
  unnest(human) %>% unnest(mouse)

#gene where names match
key_match <- key_raw %>% 
  filter(mouse == human)

key_mismatch <- key_raw %>% 
  anti_join(key_match) %>% 
  filter(!mouse %in% key_match$mouse & !human %in% key_match$human)
  # group_by(mouse) %>% 
  # summarise(human=list(human), .groups = "drop")
```

# Functions

```{r echo=FALSE}
plot_fxn <- function(dat, cell, genes, name=""){
  dat.cell <- dat %>% 
    filter(grepl(cell, group)) %>% 
    droplevels()
  dat.temp <- dat.cell %>% 
    filter(symbol %in% genes) %>% 
    droplevels()
  #Determine same direction
  wide <- dat.temp %>% 
    select(-Significance, -FDR, -ensemblID) %>% 
    group_by(group, symbol) %>% 
    summarise(logFC = mean(logFC, na.rm=TRUE), .groups = "drop") %>% 
    arrange(group) %>% 
    pivot_wider(names_from = group, values_from = logFC)
  
  pos <- rowSums(wide[-1] > 0) == ncol(wide[-1])
  neg <- rowSums(wide[-1] < 0) == ncol(wide[-1])
  na <- is.na(pos)
  #Same direction in all samples
  shared <- wide$symbol[pos | neg]
  shared <- shared[!is.na(shared)]
  #Not present in at least 1 sample
  missing <- wide$symbol[na]
  #At least 1 different direction
  diff <- wide$symbol[!wide$symbol %in% shared & !wide$symbol %in% missing] 
  
  if(length(shared) > 0){
    plot1 <- dat.temp %>% 
      filter(symbol %in% shared) %>% 
      
      ggplot(aes(reorder(symbol,logFC), logFC)) +
      geom_segment(aes(reorder(symbol,logFC), xend=symbol, y=0, yend=logFC)) +
      geom_point(size=3, aes(fill = Significance),
                 shape=21, stroke=1) +
      geom_hline(yintercept = 0) +
      
      scale_fill_brewer(palette = "RdYlBu", na.value="grey70") +
      coord_flip() +
      labs(x="", y="Log2 fold change") +
      theme_bw() +
      facet_wrap(~group, drop = FALSE, nrow = 1) +
      ggtitle(paste("Shared human and murine", cell, "responses to Mtb infection",
                    "\n", name))
  } else{
    plot1 <- plot_spacer()
  }
  
  if(length(diff) > 0){
    plot2 <- dat.temp %>% 
      filter(symbol %in% diff) %>% 
      
      ggplot(aes(reorder(symbol,logFC), logFC)) +
      geom_segment(aes(reorder(symbol,logFC), xend=symbol, y=0, yend=logFC)) +
      geom_point(size=3, aes(fill = Significance),
                 shape=21, stroke=1) +
      geom_hline(yintercept = 0) +
      
      scale_fill_brewer(palette = "RdYlBu", na.value="grey70") +
      coord_flip() +
      labs(x="", y="Log2 fold change") +
      theme_bw() +
      facet_wrap(~group, drop = FALSE, nrow = 1) +
      ggtitle(paste("Unique human and murine", cell, "responses to Mtb infection",
                    "\n", name))
  } else{
    plot2 <- plot_spacer()
  }
  
  if(length(missing) > 0){
    #Recode orthologs if available
    missing_confirm <- c()
    orth_hum <- data.frame()
    orth_mur <- data.frame()
    for(m in missing){
      wide_temp <- wide %>% filter(symbol == m)
      
      # Because of arrange in making "wide", human is always column 2 and mouse column 3
      # If missing murine data
      if(is.na(unlist(wide_temp[,3]))){
        #Look for human gene orthologs in mouse data
        orth_temp <- key_mismatch %>% filter(human == m)
        #Split genes into orthologs and truly missing
        if(nrow(orth_temp)> 0){
          orth_hum <- bind_rows(orth_temp, orth_hum)
        } else{
          missing_confirm <- c(m, missing_confirm)
        }
        #If missing human data
      } else if(is.na(unlist(wide_temp[,2]))){
        #Look for mouse gene orthologs in human data
        orth_temp <- key_mismatch %>% filter(mouse == m)
        #Split genes into orthologs and truly missing
        if(nrow(orth_temp)> 0){
          orth_mur <- bind_rows(orth_temp, orth_mur)
        } else{
          missing_confirm <- c(m, missing_confirm)
        }
      }
    }
    
    #Rename orthologs if present in data
    dat.temp.orth.rename <- data.frame()
    #human genes with mouse orthologs
    if(nrow(orth_hum) > 0){
      for(o_hum in unique(orth_hum$human)){
        orth_hum_temp <- orth_hum %>% filter(human == o_hum)
        
        if(any(orth_hum_temp$mouse %in% dat.cell[grepl("Murine", dat.cell$group),]$symbol)){
          dat.temp.orth <- dat %>% 
            filter(grepl(cell, group)) %>% 
            filter(symbol %in% orth_hum_temp$human | symbol %in% orth_hum_temp$mouse) %>% 
            droplevels()
          #Remake labels
          dat.temp.orth.rename1 <- dat.temp.orth %>% 
            filter(grepl("Murine", group)) %>% 
            mutate(orig_symbol = dat.temp.orth[grepl("Human", dat.temp.orth$group),]$symbol,
                   symbol = paste0("Human ",orig_symbol,
                                   ", Mouse ", symbol))
          #expand to repeat human for all orthologs
          dat.temp.orth.rename2 <- dat.temp.orth %>% 
            filter(grepl("Human", group)) %>% 
            rename(orig_symbol = symbol) %>% 
            full_join(select(dat.temp.orth.rename1, orig_symbol, symbol),
                      by="orig_symbol")
          #combine
          dat.temp.orth.rename <- bind_rows(dat.temp.orth.rename1, dat.temp.orth.rename2,
                                            dat.temp.orth.rename)
        } else{
          #Return gene to confirmed missing list if all orthologs also missing
          missing_confirm <- c(missing_confirm, o_hum)
        }
      }
    }
    #mouse genes with human orthologs
    ##does not occur in current genes of interset. CODE MAY NOT WORK
    if(nrow(orth_mur) > 0){
      for(o_mur in unique(orth_mur$mouse)){
        orth_mur_temp <- orth_mur %>% filter(mouse == o_mur)
        
        if(any(orth_mur_temp$human %in% dat.cell[grepl("Human", dat.cell$group),]$symbol)){
          dat.temp.orth <- dat %>% 
            filter(grepl(cell, group)) %>% 
            filter(symbol %in% orth_mur_temp$human | symbol %in% orth_mur_temp$mouse) %>% 
            droplevels()
          #Remake labels
          dat.temp.orth.rename1 <- dat.temp.orth %>% 
            filter(grepl("Human", group)) %>% 
            mutate(orig_symbol = dat.temp.orth[grepl("Mouse", dat.temp.orth$group),]$symbol,
                   symbol = paste0("Human ",symbol,
                                   ", Mouse ", orig_symbol))
          #expand to repeat human for all orthologs
          dat.temp.orth.rename2 <- dat.temp.orth %>% 
            filter(grepl("Mouse", group)) %>% 
            rename(orig_symbol = symbol) %>% 
            full_join(select(dat.temp.orth.rename1, orig_symbol, symbol),
                      by="orig_symbol")
          #combine
          dat.temp.orth.rename <- bind_rows(dat.temp.orth.rename1, dat.temp.orth.rename2,
                                            dat.temp.orth.rename)
        } else{
          #Return gene to confirmed missing list if all orthologs also missing
          missing_confirm <- c(missing_confirm, o_mur)
        }
      }
    }
    
    if(nrow(dat.temp.orth.rename) > 0){
      #plot orthologs
      plot3 <- dat.temp.orth.rename %>% 
        ggplot(aes(reorder(symbol,logFC), logFC)) +
        geom_segment(aes(reorder(symbol,logFC), xend=symbol, y=0, yend=logFC)) +
        geom_point(size=3, aes(fill = Significance),
                   shape=21, stroke=1) +
        geom_hline(yintercept = 0) +
        
        scale_fill_brewer(palette = "RdYlBu", na.value="grey70") +
        coord_flip() +
        labs(x="", y="Log2 fold change") +
        theme_bw() +
        facet_grid(orig_symbol~group, drop = FALSE, scales = "free_y", space = "free") +
        ggtitle(paste("Ortholog human or murine", cell, "responses to Mtb infection",
                      "\n", name))
    } else{
      plot3 <- plot_spacer()
    } 
    
    #Plot missing genes
    if(length(missing_confirm) > 0){
      plot4 <- dat.temp %>% 
        filter(symbol %in% missing_confirm) %>% 
        
        ggplot(aes(reorder(symbol,logFC), logFC)) +
        geom_segment(aes(reorder(symbol,logFC), xend=symbol, y=0, yend=logFC)) +
        geom_point(size=3, aes(fill = Significance),
                   shape=21, stroke=1) +
        geom_hline(yintercept = 0) +
        
        scale_fill_brewer(palette = "RdYlBu", na.value="grey70") +
        coord_flip() +
        labs(x="", y="Log2 fold change") +
        theme_bw() +
        facet_wrap(~group, drop = FALSE, nrow = 1) +
        ggtitle(paste("Missing human or murine", cell, "responses to Mtb infection",
                      "\n", name))
    } else{
      plot4 <- plot_spacer()
    } 
  }
  
  #variable ortholog plot height
  if(cell == "AM"){
    scale_fct <- 4
  }else {
    scale_fct <- 2
  }
  orth_height <- nrow(dat.temp.orth.rename)/scale_fct
  if(orth_height < 5 & orth_height > 0){ orth_height <- 5 }
  
  plot_all <- plot1/plot2/plot3/plot4 +
    plot_layout(heights = c(length(shared), length(diff), 
                            orth_height, length(missing_confirm)))
}
```

# IFN genes

```{r echo=FALSE}
#List IFN genes
ifn_genes <- unique(dat$symbol[grepl("^IFN", dat$symbol)])

#MDM
ifn_mdm_plot <- plot_fxn(dat, "MDM", ifn_genes, name="IFN genes")
#AM
ifn_am_plot <- plot_fxn(dat, "AM", ifn_genes, name="IFN genes")

#Save
ggsave("figs/gene_level/IFN_FoldChange_MDM.png", ifn_mdm_plot, 
       width = 8, height = 5)
ggsave("figs/gene_level/IFN_FoldChange_AM.png", ifn_am_plot, 
       width = 12, height = 6)
```

# IFN response genes

```{r echo=FALSE}
hum_db <- msigdbr(species = "human", category = "H") %>% 
  filter(grepl("INTERFERON", gs_name))
hum_ifna <- hum_db %>% filter(gs_name == "HALLMARK_INTERFERON_ALPHA_RESPONSE") %>% pull(gene_symbol)
hum_ifng <- hum_db %>% filter(gs_name == "HALLMARK_INTERFERON_GAMMA_RESPONSE") %>% pull(gene_symbol)

#In both terms
hum_ifnag <- intersect(hum_ifna, hum_ifng)
#Specific to 1 term
hum_ifna_spec <- hum_ifna[!hum_ifna %in% hum_ifng]
hum_ifng_spec <- hum_ifng[!hum_ifng %in% hum_ifna]
```

```{r echo=FALSE}
mur_db <- msigdbr(species = "mouse", category = "H") %>% 
  filter(grepl("INTERFERON", gs_name))
mur_ifna <- mur_db %>% filter(gs_name == "HALLMARK_INTERFERON_ALPHA_RESPONSE") %>% 
  pull(gene_symbol) %>% toupper()
mur_ifng <- mur_db %>% filter(gs_name == "HALLMARK_INTERFERON_GAMMA_RESPONSE") %>%
  pull(gene_symbol) %>% toupper()

#In both terms
mur_ifnag <- intersect(mur_ifna, mur_ifng)
#Specific to 1 term
mur_ifna_spec <- mur_ifna[!mur_ifna %in% mur_ifng]
mur_ifng_spec <- mur_ifng[!mur_ifng %in% mur_ifna]
```

Combine human and mouse IFN response genes in MSigDB Hallmark.

```{r echo=FALSE}
ALL_ifna_spec <- unique(hum_ifna_spec,mur_ifna_spec)
ALL_ifng_spec <- unique(hum_ifng_spec,mur_ifng_spec)
ALL_ifnag <- unique(hum_ifnag,mur_ifnag)
```

IFNA specific response

```{r echo=FALSE}
#MDM
ifnRa_mdm_plot <- plot_fxn(dat, "MDM", ALL_ifna_spec, name="IFNA response specific genes")
#AM
ifnRa_am_plot <- plot_fxn(dat, "AM", ALL_ifna_spec, name="IFNA response specific genes")

#Save
ggsave("figs/gene_level/IFNa_response_FoldChange_MDM.png", ifnRa_mdm_plot, 
       width = 8, height = 11)
ggsave("figs/gene_level/IFNa_response_FoldChange_AM.png", ifnRa_am_plot, 
       width = 12, height = 11)
```

IFNG specific response

```{r echo=FALSE}
#MDM
ifnRg_mdm_plot <- plot_fxn(dat, "MDM", ALL_ifng_spec, name="IFNG response specific genes")
#AM
ifnRg_am_plot <- plot_fxn(dat, "AM", ALL_ifng_spec, name="IFNG response specific genes")

#Save
ggsave("figs/gene_level/IFNg_response_FoldChange_MDM.png", ifnRg_mdm_plot, 
       width = 8, height = 22)
ggsave("figs/gene_level/IFNg_response_FoldChange_AM.png", ifnRg_am_plot, 
       width = 12, height = 22)
```

IFNA and IFNG shared response

```{r echo=FALSE}
#MDM
ifnRag_mdm_plot <- plot_fxn(dat, "MDM", ALL_ifnag, name="Shared IFNA/IFNG response genes")
#AM
ifnRag_am_plot <- plot_fxn(dat, "AM", ALL_ifnag, name="Shared IFNA/IFNG response genes")

#Save
ggsave("figs/gene_level/IFNaIFNg_response_FoldChange_MDM.png", ifnRag_mdm_plot, 
       width = 8, height = 14)
ggsave("figs/gene_level/IFNaIFNg_response_FoldChange_AM.png", ifnRag_am_plot, 
       width = 12, height = 14)
```

# R session

```{r}
# sessionInfo()
```

***
